### PR TITLE
Issue #1832 - Unit tests for meta-model structure

### LINKED ISSUE #1832: Meta-model: unit tests to verify the generated meta-model structures
### Description
Module `platform-annotation-processors` contains a class `MetaModelProcessor` which generates meta-models for domain entities.

The structure of the generated meta-models needs to be verified by a set of unit test cases. It is important to verify that a domain graph is correctly represented by the generated meta-models and can be accurately traversed.

In order to generate meta-models at runtime of the tests [`compile-testing`](https://github.com/google/compile-testing) library can be used. A set of test entities also needs to be created. For the sake of improvements in informal reasoning about the meta-model generation mechanism it is necessary that test cases have very precise and detailed names.

The whole set of possible inputs to the processor (i.e. all possible structural representations of an entity) is infinite. Therefore the following categories of entity structure are defined in order to break down the infinite set of inputs:
* An entity annotated with  `@DescTitle` (should have property `desc` metamodeled)
* An entity with sink node properties only  (all properties are metamodeled with the `PropertyMetaModel` type)
* An entity that references other entities (properties are metamodeled with the `PropertyMetaModel` + `EntityMetaModel` subclass (with lazy initialization) types)
* An entity that is a sub class of another entity (its meta-model is a sub class of the entity super class’s meta-model and it metamodels only the declared properties)
* A persistent entity (property `id` needs to be metamodeled)

### Expected outcome
A set of unit test cases with well-defined names that verify corectness of the generated meta-models' structure.

### Notes for the future
Nested entities (member classes) are currently not supported. However, this may be necessary if we start supporting components that are declared as nested types.

### PR DESCRIPTION
**Pull request by homedirectory**
Resolved #1832.
Did I set the base branch correctly? It's `migration-to-java-11-with-EQL3`, right?

### CODE REVIEW

File: platform-annotation-processors/src/main/java/ua/com/fielden/platform/processors/metamodel/utils/EntityFinder.java [hunk 1]
```diff
@@ -46,6 +45,20 @@ private EntityFinder() {}
...
     public static final Class<?> ROOT_ENTITY_CLASS = AbstractEntity.class;
 
+    /**
+     * Finds an entity described by {@code entityClass} in the environment provided with {@code elementUtils}.
+     * @param entityClass
+     * @param elementUtils
+     * @return {@link EntityElement} wrapped in an {@link Optional} if found, else an empty optional
+     */
+    public static Optional<EntityElement> findEntity(final Class<? extends AbstractEntity<?>> entityClass, final Elements elementUtils) {
+        final Optional<TypeElement> typeElement = Optional.ofNullable(elementUtils.getTypeElement(entityClass.getCanonicalName()));
```
Comment (01es) [inline platform-annotation-processors/src/main/java/ua/com/fielden/platform/processors/metamodel/utils/EntityFinder.java:55]: How about just:
```Java
return Optional.ofNullable(elementUtils.getTypeElement(entityClass.getCanonicalName())).map(te -> new EntityElement(typeElement.get(), elementUtils));
```

File: platform-annotation-processors/src/main/java/ua/com/fielden/platform/processors/metamodel/utils/EntityFinder.java [hunk 2]
```diff
@@ -233,8 +246,15 @@ public static boolean isPropertyOfDomainEntityType(final PropertyElement propEle
      * @return
      */
     public static boolean isEntityThatNeedsMetaModel(final TypeElement element) {
-        return EntityFinder.isEntityType(element) && 
-               (element.getAnnotation(MapEntityTo.class) != null || element.getAnnotation(DomainEntity.class) != null || element.getAnnotation(WithMetaModel.class) != null);
+        if (!EntityFinder.isEntityType(element)) {
+            return false;
+        }
+        for (final Class<? extends Annotation> annotClass: ANNOTATIONS_THAT_TRIGGER_META_MODEL_GENERATION)  {
```
Comment (01es) [inline platform-annotation-processors/src/main/java/ua/com/fielden/platform/processors/metamodel/utils/EntityFinder.java:252]: Try to convert this to a `return` (i.e., a single expression) by employing more functional constructs.

File: platform-annotation-processors/src/main/java/ua/com/fielden/platform/processors/metamodel/utils/MetaModelFinder.java [hunk 3]
```diff
@@ -83,23 +113,40 @@ public static Set<MetaModelElement> findReferencedMetaModels(final MetaModelElem
...
+    
+    /**
+     * Attempts to find a meta-model for a given entity.
+     * @param entityElement
+     * @param elementUtils
+     * @return {@link MetaModelElement} instance wrapped into {@link Optional} if found, else empty optional
+     */
+    public static Optional<MetaModelElement> findMetaModelForEntity(final EntityElement entityElement, final Elements elementUtils) {
+        final MetaModelConcept mmc = new MetaModelConcept(entityElement);
+        final Optional<TypeElement> maybeMmte = Optional.ofNullable(elementUtils.getTypeElement(mmc.getQualifiedName()));
```
Comment (01es) [inline platform-annotation-processors/src/main/java/ua/com/fielden/platform/processors/metamodel/utils/MetaModelFinder.java:146]: Try to rewrite this into a single `return` statement using `.map` as per earlier example.

**Comment by homedirectory** [comment]
This branch was merged into Issue-#1855. Closing this pull request.

**Pull request closed without merging**